Adopt C++20 Concepts in JavaScriptCore/jit

iidmsa opened the pull request:
#### 6157d21f7a613e8f47ebce0bbeb2b13891cc95ed
<pre>
Adopt C++20 Concepts in JavaScriptCore/jit
<a href="https://bugs.webkit.org/show_bug.cgi?id=303249">https://bugs.webkit.org/show_bug.cgi?id=303249</a>
<a href="https://rdar.apple.com/165542109">rdar://165542109</a>

Reviewed by Justin Michaud and Darin Adler.

* Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
(JSC::AssemblyHelpers::CopySpooler::store):
(JSC::AssemblyHelpers::CopySpooler::finalize):
* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::pokeForArgument):
(JSC::CCallHelpers::setupArgumentsImpl):
(JSC::CCallHelpers::setupArgumentsEntryImpl):
(JSC::CCallHelpers::std::is_integral&lt;CURRENT_ARGUMENT_TYPE&gt;::value): Deleted.
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
* Source/JavaScriptCore/jit/GPRInfo.h:
(JSC::PreferredArgumentImpl::sizeOfArg):
(JSC::PreferredArgumentImpl::preferredArgumentJSR):
(JSC::PreferredArgumentImpl::preferredArgumentGPR):
(JSC::preferredArgumentGPR):
(JSC::preferredArgumentJSR):
* Source/JavaScriptCore/jit/JITCompilationKey.h:
(JSC::JITCompilationKey::hash const):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::compileOpStrictEq):
(JSC::JIT::compileOpStrictEqJump):
* Source/WTF/wtf/StdLibExtras.h:

Canonical link: <a href="https://commits.webkit.org/303871@main">https://commits.webkit.org/303871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ca00d625f203d9e2c8833a3b716792b0bb7bc73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133628 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6133 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141191 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85682 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102221 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69579 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136575 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4745 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83021 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4624 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2198 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125705 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113766 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37910 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143837 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132142 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5805 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38491 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110604 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4969 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110788 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28139 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4441 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116049 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59554 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5857 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34360 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165105 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5719 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69315 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43140 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5949 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5812 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/jit/CCallHelpers.h:
```diff
@@ -531,15 +531,13 @@ class CCallHelpers : public AssemblyHelpers {
-        std::is_base_of<TrustedImm, Arg>::value
-        || std::is_convertible<Arg, TrustedImm>::value> // We have this since DFGSpeculativeJIT has it's own implementation of TrustedImmPtr
-    setupArgumentsImpl(ArgCollection<numGPRArgs, numGPRSources, numFPRArgs, numFPRSources, numCrossSources, extraGPRArgs, nonArgGPRs, extraPoke> argSourceRegs, Arg arg, Args... args)
+        requires WTF::LooselyDerivedFrom<Arg, TrustedImm>
```
Doesn't seem great to remove the comment about why the requirements here are a little unusual. Also, I'm not sure the term "loosely" is unambiguous enough to make it clear when we should use this and when we should use `derived_from`, and lacking a name that makes it obvious, there's also no comment to explain.

iidmsa replied:
Will add back the comment about requirements. Will resort to `DerivedFromOrConvertibleTo`.

darinadler commented on Source/JavaScriptCore/jit/GPRInfo.h:
```diff
@@ -955,10 +955,14 @@ class NoOverlapImpl {
 constexpr bool noOverlap(Args... args) { return NoOverlapImpl::entry(args...); }
 
+template<typename OperationType, unsigned ArgNum>
+concept HasArgument = (FunctionTraits<OperationType>::arity > ArgNum);
```
Not sure the name makes it clear this would take an argument number. Also seems more natural to use `size_t` rather than `unsigned`.

iidmsa replied:
Will resort to `HasNthArgument` and `size_t`.

darinadler commented on Source/JavaScriptCore/jit/CCallHelpers.h:
```diff
@@ -531,15 +531,13 @@ class CCallHelpers : public AssemblyHelpers {
-        std::is_base_of<TrustedImm, Arg>::value
-        || std::is_convertible<Arg, TrustedImm>::value> // We have this since DFGSpeculativeJIT has it's own implementation of TrustedImmPtr
-    setupArgumentsImpl(ArgCollection<numGPRArgs, numGPRSources, numFPRArgs, numFPRSources, numCrossSources, extraGPRArgs, nonArgGPRs, extraPoke> argSourceRegs, Arg arg, Args... args)
+        requires WTF::DerivedFromOrConvertibleTo<Arg, TrustedImm> // We have this since DFGSpeculativeJIT has it's own implementation of TrustedImmPtr
```
```suggestion
        requires WTF::DerivedFromOrConvertibleTo<Arg, TrustedImm> // DerivedFromOrConvertibleTo instead of derived_from since DFGSpeculativeJIT has its own implementation of TrustedImmPtr

darinadler commented on Source/JavaScriptCore/jit/GPRInfo.h:
```diff
@@ -955,16 +955,20 @@ class NoOverlapImpl {
 constexpr bool noOverlap(Args... args) { return NoOverlapImpl::entry(args...); }
 
+template<typename OperationType, size_t ArgNum>
+concept HasNthArgument = (FunctionTraits<OperationType>::arity > ArgNum);
```
```suggestion
concept HasNthArgument = FunctionTraits<OperationType>::arity > ArgNum;
```

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -535,6 +535,9 @@ template<typename T> concept HasSwitchOn = requires(T t) {
 };
 
+template<typename Derived, typename Base>
+concept DerivedFromOrConvertibleTo = std::is_base_of_v<Base, Derived> || std::is_convertible_v<Derived, Base>;
```
Why does this use `is_base_of_v` instead of `derived_from`?

iidmsa replied:
`std::derived_from` only accepts public inheritance, but we need to accept private inheritance as well (which `std::is_base_of_v` does). This should handle the case for `DFGSpeculativeJIT` custom `TrustedImmPtr` implementation.

iidmsa committed:
Adopt C++20 Concepts in JavaScriptCore/jit
https://bugs.webkit.org/show_bug.cgi?id=303249
rdar://165542109

Reviewed by Justin Michaud and Darin Adler.

* Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
(JSC::AssemblyHelpers::CopySpooler::store):
(JSC::AssemblyHelpers::CopySpooler::finalize):
* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::pokeForArgument):
(JSC::CCallHelpers::setupArgumentsImpl):
(JSC::CCallHelpers::setupArgumentsEntryImpl):
(JSC::CCallHelpers::std::is_integral<CURRENT_ARGUMENT_TYPE>::value): Deleted.
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
* Source/JavaScriptCore/jit/GPRInfo.h:
(JSC::PreferredArgumentImpl::sizeOfArg):
(JSC::PreferredArgumentImpl::preferredArgumentJSR):
(JSC::PreferredArgumentImpl::preferredArgumentGPR):
(JSC::preferredArgumentGPR):
(JSC::preferredArgumentJSR):
* Source/JavaScriptCore/jit/JITCompilationKey.h:
(JSC::JITCompilationKey::hash const):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::compileOpStrictEq):
(JSC::JIT::compileOpStrictEqJump):
* Source/WTF/wtf/StdLibExtras.h:

Canonical link: https://commits.webkit.org/303871@main
